packages/jdk.rb

```diff
@@ -1,43 +1,51 @@
 require 'package'
 
 class Jdk < Package
-  description 'The JDK is a development environment for building applications, applets, and components using the Java programming language.'
-  homepage 'https://www.oracle.com/java/technologies/javase/javase-jdk8-downloads.html'
-  version '8u333-18.0.1.1'
+  description 'The Oracle JDK is a development environment for building applications, applets, and components using the Java programming language.'
+  homepage 'https://www.oracle.com/java'
   license 'Oracle-BCLA-JavaSE'
   compatibility 'all'
 
+  # full version number extracted from jdk*.rb recipes
+  @avail_jdk_ver = [8, 11, 17, 18].to_h do |jdk_majver|
+    pkg = Package.load_package("#{CREW_PACKAGES_PATH}/jdk#{jdk_majver}.rb")
+
+    [jdk_majver, pkg.version]
+  end
+
+  version "#{@avail_jdk_ver.to_a[0][1]}-#{@avail_jdk_ver.to_a[-1][1]}"
+
   is_fake
 
   def self.preflight
-    jdkver = `java -version 2>&1 | head -1 | cut -d'"' -f2`.chomp
-    unless jdkver.empty? || jdkver.include?('No such file or directory') || jdkver.include?('not found')
-      majver = `java -version 2>&1 | head -1 | cut -d'"' -f2 | cut -d'.' -f1`.chomp
-      majver = '8' if majver == '1'
-      puts "Package jdk#{majver} already installed.".lightgreen
-      abort "Enter `crew remove jdk#{majver} && crew install jdk` to install a different version."
-    end
-    puts
-    puts 'Select version:'
-    puts ' 8 = JDK 8u333'
-    if ARCH == 'x86_64'
-      puts '11 = JDK 11.0.16.1'
-      puts '17 = JDK 17.0.3.1'
-      puts '18 = JDK 18.0.1.1'
-    end
-    puts ' 0 = Cancel'
-
-    while version = $stdin.gets.chomp
-      case version
-      when '8', '11', '17', '18'
-        depends_on "jdk#{version}"
-        break
-      when '0'
-        abort
-        break
-      else
-        puts "\nPlease select from one of the options or enter 0 to cancel."
+    if ARGV.include?('install')
+      jdk_exec = File.join(CREW_PREFIX, 'bin', 'java')
+
+      if File.exist?(jdk_exec)
+        jdk_ver_str   = `#{jdk_exec} -version 2>&1`
+        jdk_ver       = jdk_ver_str[/version "(.+?)"/, 1]
+        jdk_major_ver = jdk_ver.match?(/^1.8/) ? '8' : jdk_ver.partition('.')[0]
+
+        is_openjdk   = jdk_ver_str.include?('openjdk')
+        pkg_branding = is_openjdk ? 'OpenJDK' : 'Oracle JDK'
+        pkg_prefix   = is_openjdk ? 'openjdk' : 'jdk'
+
+        abort <<~EOT.yellow
+
+          #{pkg_branding} #{jdk_ver} installed.
+
+          Run "crew remove #{pkg_prefix}#{jdk_major_ver}; crew install #{name}" to install another version of JDK
+        EOT
       end
+
+      options = [
+        { value: 'jdk8',  description: "Oracle JDK #{@avail_jdk_ver[8]}"  },
+        { value: 'jdk11', description: "Oracle JDK #{@avail_jdk_ver[11]}" },
+        { value: 'jdk17', description: "Oracle JDK #{@avail_jdk_ver[17]}" },
+        { value: 'jdk18', description: "Oracle JDK #{@avail_jdk_ver[18]}" }
+      ]
```

Comment:
Build `options` based on the package file versions found in `#{CREW_PACKAGES_PATH}/jdk*.rb`?

Reply:
Change made

Reply:
Nice!